src/main/fc/gps_lap_timer.c

```diff
@@ -0,0 +1,150 @@
+    gpsLapTimerData.gateLocation.lon  = gpsLapTimerConfig()->gateLon;
+    gpsLapTimerData.numberOfSetReadings = 0;
+    gpsLapTimerData.bestLapTime = 0;
+    gpsLapTimerData.best3Consec = 0;
+    for (uint8_t i = 0; i < MAX_N_RECORDED_PREVIOUS_LAPS; i++) {
+        gpsLapTimerData.previousLaps[i] = 0;
+    }
+    gpsLapTimerData.timerRunning = false;
+    gpsLapTimerData.timeOfLastLap = 0;
+    gpsLapTimerData.numberOfLapsRecorded = 0;
+}
+
+void gpsLapTimerStartSetGate(void)
+{
+    settingGate = true;
+    gateSetLatReadings = 0;
+    gateSetLonReadings = 0;
+    gpsLapTimerData.numberOfSetReadings = 0;
+}
+
+void gpsLapTimerProcessSettingGate(void)
+{
+    if (gpsLapTimerData.numberOfSetReadings < MAX_GATE_SET_READINGS){
+        gateSetLatReadings += gpsSol.llh.lat;
+        gateSetLonReadings += gpsSol.llh.lon;
+        gpsLapTimerData.numberOfSetReadings++;
+    }
+    gpsLapTimerData.gateLocation.lat = gateSetLatReadings / gpsLapTimerData.numberOfSetReadings;
+    gpsLapTimerData.gateLocation.lon = gateSetLonReadings / gpsLapTimerData.numberOfSetReadings;
+}
+
+void gpsLapTimerEndSetGate(void)
+{
+    settingGate = false;
+    gpsLapTimerConfigMutable()->gateLat = gpsLapTimerData.gateLocation.lat;
+    gpsLapTimerConfigMutable()->gateLon = gpsLapTimerData.gateLocation.lon;
+}
+
+void gpsLapTimerNewGpsData(void)
+{
+    GPS_distance_cm_bearing(&gpsSol.llh.lat, &gpsSol.llh.lon, &gpsLapTimerData.gateLocation.lat, &gpsLapTimerData.gateLocation.lon, &gpsLapTimerData.distToPointCM, &gpsLapTimerData.dirToPoint);
+    
+    if (settingGate) {
+        gpsLapTimerProcessSettingGate();
+        return;
+    }
+
+    // Reset all lap timer data if reset mode is active
+    if (IS_RC_MODE_ACTIVE(BOXLAPTIMERRESET)) {
+        gpsLapTimerInit();
+        return;
+    }
+
+    // Current lap time is at least the min lap timer or timer not running, so we need to get ready to record a gate crossing
+    if (!gpsLapTimerData.timerRunning || gpsSol.time - gpsLapTimerData.timeOfLastLap > (gpsLapTimerConfig()->minimumLapTimeSeconds * 1000)) {
+        // Within radius of gate, record the closest point we get before leaving
+        if (gpsLapTimerData.distToPointCM < (gpsLapTimerConfig()->gateToleranceM * 100)) {
+            // Either just entered the circle or were already in circle but are the closest we've been to the center this lap
+            if (!wasInCircle || gpsLapTimerData.distToPointCM < minDistance) {
+                // Track the closest we've been to the center
+                minDistance = gpsLapTimerData.distToPointCM;
+                // Track the time we were the closest to the center, which will be used to determine the actual lap time
+                minDistanceTime = gpsSol.time;
+            }
+            wasInCircle = true;
+        } else {
+            // Just left the circle, so record the lap time
+            if (wasInCircle) {
+                // Not the first time through the gate
```

Comment:
If you have a compass you could dial in the heading of the gate directly through the CMS menu or CLI

Reply:
@SteveCEvans IIRC gate time is taken at closest point to gate (minimum distance is tracked, with timestamp). So actual gate size should not matter.

Reply:
An exit direction could help by letting us make the circle bigger and not be caught by other parts of the track. Since it's using the closest point to the gate waypoint, I think that's the most accurate we can be without correcting the gate waypoint. I'm sure there is a lot more we could do with this in the future. Split times and auto-correcting the start/finish point are two things I've thought of.